tools/tpm2_flushcontext.c

```diff
@@ -0,0 +1,183 @@
+            return "saved session";
+    }
+
+    return "invalid";
+}
+
+static TSS2_RC
+get_capability_handles(TSS2_SYS_CONTEXT *sapi_ctx, UINT32 property,
+                       TPMS_CAPABILITY_DATA *capability_data) {
+
+    TSS2_RC rc;
+    TPMI_YES_NO more_data;
+
+    rc = Tss2_Sys_GetCapability(sapi_ctx, NULL, TPM2_CAP_HANDLES, property,
+                                TPM2_MAX_CAP_HANDLES, &more_data,
+                                capability_data,
+                                NULL);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERR("Failed to GetCapability: capability: 0x%x, property: 0x%x, "
+                "TSS2_RC: 0x%x", TPM2_CAP_HANDLES, property, rc);
+    } else if (more_data == YES) {
+        LOG_WARN("More data to be queried: capability: 0x%x, property: "
+                 "0x%x", TPM2_CAP_HANDLES, property);
+    }
+
+    return rc;
+}
+
+static int flush_contexts(TSS2_SYS_CONTEXT *sapi_context, TPM2_HANDLE handles[],
+                          UINT32 count) {
+
+    UINT32 i;
+
+    for (i = 0; i < count; ++i) {
+        TPM2_RC rc;
+
+        rc = Tss2_Sys_FlushContext(sapi_context, handles[i]);
+        if (rc != TPM2_RC_SUCCESS) {
+            LOG_ERR("Failed Flush Context for %s handle 0x%x, TSS2_RC: 0x%x",
+                    get_property_name(handles[i]), handles[i], rc);
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+static bool on_option(char key, char *value) {
+
+    bool result;
+
+    switch (key) {
+    case 'H':
+        result = tpm2_util_string_to_uint32(value, &ctx.objectHandle);
+        if (!result) {
+            return false;
+        }
+        ctx.flags.H = 1;
+        break;
+    case 't':
+        ctx.flags.t = 1;
+        break;
+    case 'l':
+        ctx.flags.l = 1;
+        break;
+    case 's':
+        ctx.flags.s = 1;
+        break;
+    }
+
+    return true;
+}
+
+bool tpm2_tool_onstart(tpm2_options **opts) {
+
+    static const struct option topts[] = {
+        { "handle",           no_argument, NULL, 'H' },
+        { "transient-object", no_argument, NULL, 't' },
+        { "loaded-session",   no_argument, NULL, 'l' },
+        { "saved-session",    no_argument, NULL, 's' },
+    };
+
+    *opts = tpm2_options_new("H:tls", ARRAY_LEN(topts), topts,
+                             on_option, NULL);
+
+    return *opts != NULL;
+}
+
+int tpm2_tool_onrun(TSS2_SYS_CONTEXT *sapi_context, tpm2_option_flags flags) {
+
+    UNUSED(flags);
+
+    if (!(ctx.flags.H || ctx.flags.t || ctx.flags.l || ctx.flags.s)) {
+        return false;
+    }
+
+    TPMS_CAPABILITY_DATA capability_data;// = TPMS_CAPABILITY_DATA_EMPTY_INIT;
+    TPML_HANDLE *handles = &capability_data.data.handles;
+    UINT32 property;
+
+    if (ctx.flags.t) {
+        property = TPM2_TRANSIENT_FIRST;
+    } else if (ctx.flags.l) {
+        property = TPM2_LOADED_SESSION_FIRST;
+    } else if (ctx.flags.s) {
+        property = TPM2_ACTIVE_SESSION_FIRST;
```

Comment:
Since there are mutually exclusive, I would just have a UINT32 property field in the struct tpm_flush_context_ctx and set it to TPM2_*_FIRST in the on_option() callback. I don't see the need for the flags.